src/main/java/spoon/support/reflect/code/CtCaseImpl.java

```diff
@@ -56,11 +56,13 @@ public void accept(CtVisitor visitor) {
 
 	@Override
 	public <T extends CtCase<E>> T setCaseExpression(CtExpression<E> caseExpression) {
-		if (caseExpression != null) {
-			caseExpression.setParent(this);
+		if (caseExpression == null) {
+			this.caseExpressions = emptyList();
+			return (T) this;
```

Comment:
The architecture test that fails is that each setter should ensure that the set child has the correct parent. It's not sophisticated enough to figure out that the parent is set later. Something like this should do as a workaround:

```suggestion
		if (caseExpression != null) {
			// setting the parent here is redundant as it's later set in addCaseExpression, but we need to set it
                        // here to appease the architecture tests that check for parent consistency
			caseExpression.setParent(this);

Reply:
Perhaps @monperrus has a less dirty workaround?

Reply:
> setting the parent here is redundant as it's later set in addCaseExpression,

That's exactly why I didn't write that statement here.

> but we need to set it here to appease the architecture tests that check for

Why doesn't [this setter](https://github.com/INRIA/spoon/blob/master/src/main/java/spoon/support/reflect/code/CtCaseImpl.java#L73) fail in that case? This also sets the parent later in `addExpression`.

Reply:
The test only looks for setters where the argument is a subtype of `CtElement`. The one you linked takes a list.